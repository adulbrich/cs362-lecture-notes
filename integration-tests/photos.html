<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Photos</title>
    <script type="module" src="photos.js" defer></script>
    <script src="https://cdn.jsdelivr.net/npm/@tailwindcss/browser@4"></script>
  </head>
  <body class="bg-gray-50 min-h-screen">
    <div class="max-w-md mx-auto bg-white rounded-xl shadow-md p-6 m-6">
      <h1 class="text-2xl font-bold text-gray-800 mb-6">Photo Gallery</h1>
      <form id="add-photo-form" class="mb-6">
        <div class="flex flex-col mb-6">
          <label class="font-medium text-gray-700">
            Photo URL
            <input
              id="url"
              name="url"
              class="block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm"
              placeholder="https://example.com/image.jpg"
            />
          </label>
        </div>
        <div class="flex flex-col mb-6">
          <label class="font-medium text-gray-700">
            Caption
            <input
              id="caption"
              name="caption"
              class="block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm"
              placeholder="Enter a caption"
            />
          </label>
        </div>
        <button
          type="submit"
          class="w-full flex justify-center py-2 px-4 mb-6 border border-transparent rounded-md shadow-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700"
        >
          Add photo
        </button>
      </form>
      <ul id="photo-card-list" class="flex flex-col gap-6 justify-center"></ul>
    </div>
  </body>
</html>
