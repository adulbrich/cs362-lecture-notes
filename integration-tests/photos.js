function insertNewPhotoCard(url, caption) {
  const photoItem = document.createElement("li");
  photoItem.classList.add("justify-items-center");

  const photoImg = document.createElement("img");
  photoImg.src = url;
  photoImg.alt = caption;

  const captionP = document.createElement("p");
  captionP.textContent = caption;

  photoItem.append(photoImg, captionP);
  const photoContainer = document.getElementById("photo-card-list");
  photoContainer.appendChild(photoItem);
}

export function setupPhotoForm() {
  const form = document.getElementById("add-photo-form");
  const urlInput = document.getElementById("url");
  const captionInput = document.getElementById("caption");

  form.addEventListener("submit", function (event) {
    event.preventDefault();
    const url = urlInput.value;
    const caption = captionInput.value;
    if (url && caption) {
      insertNewPhotoCard(url, caption);
      urlInput.value = "";
      captionInput.value = "";
    }
  });
}

document.addEventListener("DOMContentLoaded", setupPhotoForm);
