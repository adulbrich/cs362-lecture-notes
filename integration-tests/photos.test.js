/**
 * @vitest-environment jsdom
 */

import fs from "fs";
import path from "path";
import { describe, test, expect, beforeEach } from "vitest";
import { screen } from "@testing-library/dom";
import "@testing-library/jest-dom/vitest";
import { userEvent } from "@testing-library/user-event";
import { setupPhotoForm } from "./photos.js";

describe("Photos Application", () => {
  beforeEach(() => {
    const htmlPath = path.resolve(__dirname, "./photos.html");
    const htmlContent = fs.readFileSync(htmlPath, "utf-8");
    document.body.innerHTML = htmlContent;
    setupPhotoForm();
  });

  test("should insert a new photo card when the form is submitted", async () => {
    const urlInput = screen.getByLabelText(/url/i);
    const captionInput = screen.getByLabelText(/caption/i);
    const submitButton = screen.getByRole("button", { name: /add photo/i });

    await userEvent.type(urlInput, "https://picsum.photos/200/300");
    await userEvent.type(captionInput, "Lorem Picsum");
    await userEvent.click(submitButton);

    const addedPhoto = screen.getByRole("img", { name: "Lorem Picsum" });
    expect(addedPhoto).toBeVisible();
    expect(addedPhoto).toHaveAttribute("src", "https://picsum.photos/200/300");
    expect(addedPhoto).toHaveAttribute("alt", "Lorem Picsum");

    const caption = screen.getByText("Lorem Picsum");
    expect(caption).toBeVisible();
  });

  test("should clear form fields after successful submission", async () => {
    const urlInput = screen.getByLabelText(/url/i);
    const captionInput = screen.getByLabelText(/caption/i);
    const submitButton = screen.getByRole("button", { name: /add photo/i });

    await userEvent.type(urlInput, "https://picsum.photos/200/300");
    await userEvent.type(captionInput, "Lorem Picsum");
    await userEvent.click(submitButton);

    expect(urlInput).toHaveDisplayValue("");
    expect(captionInput).toHaveDisplayValue("");
  });

  test("should not add photo card or clear form when missing url", async () => {
    const captionInput = screen.getByLabelText(/caption/i);
    const submitButton = screen.getByRole("button", { name: /add photo/i });
    const photoCardList = screen.getByRole("list");

    await userEvent.type(captionInput, "Lonely Caption :(");
    await userEvent.click(submitButton);

    expect(photoCardList).toBeEmptyDOMElement();
    expect(captionInput).toHaveDisplayValue("Lonely Caption :(");
  });

  test("should not add photo card or clear form when missing caption", async () => {
    const urlInput = screen.getByLabelText(/url/i);
    const submitButton = screen.getByRole("button", { name: /add photo/i });

    await userEvent.type(urlInput, "https://picsum.photos/200/300");
    await userEvent.click(submitButton);

    const potentialImage = screen.queryByRole("img", {
      src: "https://picsum.photos/200/300",
    });
    expect(potentialImage).not.toBeInTheDocument();

    expect(urlInput).toHaveDisplayValue("https://picsum.photos/200/300");
  });

  test("should insert multiple photo cards when submitted sequentially", async () => {
    const urlInput = screen.getByLabelText(/url/i);
    const captionInput = screen.getByLabelText(/caption/i);
    const submitButton = screen.getByRole("button", { name: /add photo/i });
    const photoCardList = screen.getByRole("list");

    await userEvent.type(urlInput, "https://picsum.photos/200/300");
    await userEvent.type(captionInput, "Lorem Picsum 200x300");
    await userEvent.click(submitButton);

    await userEvent.type(urlInput, "https://picsum.photos/300/400");
    await userEvent.type(captionInput, "Lorem Picsum 300x400");
    await userEvent.click(submitButton);

    expect(photoCardList).toMatchSnapshot();
  });
});
