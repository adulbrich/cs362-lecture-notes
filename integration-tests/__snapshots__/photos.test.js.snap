// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Photos Application > should insert multiple photo cards when submitted sequentially 1`] = `
<ul
  class="flex flex-col gap-6 justify-center"
  id="photo-card-list"
>
  <li
    class="justify-items-center"
  >
    <img
      alt="Lorem Picsum 200x300"
      src="https://picsum.photos/200/300"
    />
    <p>
      Lorem Picsum 200x300
    </p>
  </li>
  <li
    class="justify-items-center"
  >
    <img
      alt="Lorem Picsum 300x400"
      src="https://picsum.photos/300/400"
    />
    <p>
      Lorem Picsum 300x400
    </p>
  </li>
</ul>
`;
